crates/sui-types/src/multisig.rs

```diff
@@ -0,0 +1,316 @@
+                    .ok_or(SuiError::InvalidSignature {
+                        error: "Invalid public keys index".to_string(),
+                    })?;
+            let res = match sig {
+                CompressedSignature::Ed25519(s) => {
+                    let pk = Ed25519PublicKey::from_bytes(pk_map.0.as_ref()).map_err(|_| {
+                        SuiError::InvalidSignature {
+                            error: "Invalid public key".to_string(),
+                        }
+                    })?;
+                    pk.verify(&message, s)
+                }
+                CompressedSignature::Secp256k1(s) => {
+                    let pk = Secp256k1PublicKey::from_bytes(pk_map.0.as_ref()).map_err(|_| {
+                        SuiError::InvalidSignature {
+                            error: "Invalid public key".to_string(),
+                        }
+                    })?;
+                    pk.verify(&message, s)
+                }
+                CompressedSignature::Secp256r1(s) => {
+                    let pk = Secp256r1PublicKey::from_bytes(pk_map.0.as_ref()).map_err(|_| {
+                        SuiError::InvalidSignature {
+                            error: "Invalid public key".to_string(),
+                        }
+                    })?;
+                    pk.verify(&message, s)
+                }
+            };
+            if res.is_ok() {
+                weight_sum += pk_map.1 as u16;
+            }
+        }
+
+        if weight_sum >= self.multi_pk.threshold {
+            Ok(())
+        } else {
+            Err(SuiError::InvalidSignature {
+                error: format!("Insufficient weight {:?}", weight_sum),
+            })
+        }
+    }
+}
+
+impl MultiSignature {
+    /// This combines a list of [enum Signature] `flag || signature || pk` to a multisignature.
+    pub fn combine(full_sigs: Vec<Signature>, multi_pk: MultiPublicKey) -> Result<Self, SuiError> {
+        if full_sigs.len() > multi_pk.pk_map.len()
+            || multi_pk.pk_map.len() > MAX_SIGNER_IN_MULTISIG
+            || full_sigs.is_empty()
+            || multi_pk.pk_map.is_empty()
+        {
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid number of signatures".to_string(),
+            });
+        }
+        let mut bitmap = RoaringBitmap::new();
+        let mut sigs = Vec::new();
+        full_sigs.iter().for_each(|s| {
+            // TODO(joyqvq): Remove unwrap here.
+            bitmap.insert(multi_pk.get_index(s.to_public_key().unwrap()).unwrap());
+            sigs.push(s.to_compressed().unwrap());
+        });
+
+        Ok(MultiSignature {
+            sigs,
+            bitmap,
+            multi_pk,
+            bytes: OnceCell::new(),
+        })
+    }
+}
+
+/// The struct that contains the public key used for authenticating a multisig.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
+pub struct MultiPublicKey {
```

Comment:
nit, and obviously very subjective - I view "MultiSig" as the scheme name, like ECDSA or EdDSA, and so MultiSigPublicKey sounds clearer to me here.